{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"authorship_tag":"ABX9TyMQERKg6WGBOzkP7Br/lWEt"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"},"widgets":{"application/vnd.jupyter.widget-state+json":{"21a6391881964522adcb3879ea507e9c":{"model_module":"@jupyter-widgets/controls","model_name":"HBoxModel","model_module_version":"1.5.0","state":{"_dom_classes":[],"_model_module":"@jupyter-widgets/controls","_model_module_version":"1.5.0","_model_name":"HBoxModel","_view_count":null,"_view_module":"@jupyter-widgets/controls","_view_module_version":"1.5.0","_view_name":"HBoxView","box_style":"","children":["IPY_MODEL_6214dfffeb5c4ec7a8f9f9182e9e7603","IPY_MODEL_39ffd7cecf664f5ebe61fa8047b90d44","IPY_MODEL_838d93b78b7c4249afe9c761886ccc86"],"layout":"IPY_MODEL_a18d2efb1ef24c8e9060bc714b288ef0"}},"6214dfffeb5c4ec7a8f9f9182e9e7603":{"model_module":"@jupyter-widgets/controls","model_name":"HTMLModel","model_module_version":"1.5.0","state":{"_dom_classes":[],"_model_module":"@jupyter-widgets/controls","_model_module_version":"1.5.0","_model_name":"HTMLModel","_view_count":null,"_view_module":"@jupyter-widgets/controls","_view_module_version":"1.5.0","_view_name":"HTMLView","description":"","description_tooltip":null,"layout":"IPY_MODEL_2732175248a1418a9fab0f0cd845ecb7","placeholder":"​","style":"IPY_MODEL_6d1c7f3a39794b70bc22b0f823d53825","value":"Evaluating: 100%"}},"39ffd7cecf664f5ebe61fa8047b90d44":{"model_module":"@jupyter-widgets/controls","model_name":"FloatProgressModel","model_module_version":"1.5.0","state":{"_dom_classes":[],"_model_module":"@jupyter-widgets/controls","_model_module_version":"1.5.0","_model_name":"FloatProgressModel","_view_count":null,"_view_module":"@jupyter-widgets/controls","_view_module_version":"1.5.0","_view_name":"ProgressView","bar_style":"success","description":"","description_tooltip":null,"layout":"IPY_MODEL_50049a3cc27146158d3f70c37e297568","max":4,"min":0,"orientation":"horizontal","style":"IPY_MODEL_5ad7fdb1a757485988c0ae1082663997","value":4}},"838d93b78b7c4249afe9c761886ccc86":{"model_module":"@jupyter-widgets/controls","model_name":"HTMLModel","model_module_version":"1.5.0","state":{"_dom_classes":[],"_model_module":"@jupyter-widgets/controls","_model_module_version":"1.5.0","_model_name":"HTMLModel","_view_count":null,"_view_module":"@jupyter-widgets/controls","_view_module_version":"1.5.0","_view_name":"HTMLView","description":"","description_tooltip":null,"layout":"IPY_MODEL_7f9a1321cb4a4882854ad29e7d2c4fbc","placeholder":"​","style":"IPY_MODEL_40e3ded372eb4f268ebc13c6b2c3b26c","value":" 4/4 [03:04&lt;00:00, 34.95s/it]"}},"a18d2efb1ef24c8e9060bc714b288ef0":{"model_module":"@jupyter-widgets/base","model_name":"LayoutModel","model_module_version":"1.2.0","state":{"_model_module":"@jupyter-widgets/base","_model_module_version":"1.2.0","_model_name":"LayoutModel","_view_count":null,"_view_module":"@jupyter-widgets/base","_view_module_version":"1.2.0","_view_name":"LayoutView","align_content":null,"align_items":null,"align_self":null,"border":null,"bottom":null,"display":null,"flex":null,"flex_flow":null,"grid_area":null,"grid_auto_columns":null,"grid_auto_flow":null,"grid_auto_rows":null,"grid_column":null,"grid_gap":null,"grid_row":null,"grid_template_areas":null,"grid_template_columns":null,"grid_template_rows":null,"height":null,"justify_content":null,"justify_items":null,"left":null,"margin":null,"max_height":null,"max_width":null,"min_height":null,"min_width":null,"object_fit":null,"object_position":null,"order":null,"overflow":null,"overflow_x":null,"overflow_y":null,"padding":null,"right":null,"top":null,"visibility":null,"width":null}},"2732175248a1418a9fab0f0cd845ecb7":{"model_module":"@jupyter-widgets/base","model_name":"LayoutModel","model_module_version":"1.2.0","state":{"_model_module":"@jupyter-widgets/base","_model_module_version":"1.2.0","_model_name":"LayoutModel","_view_count":null,"_view_module":"@jupyter-widgets/base","_view_module_version":"1.2.0","_view_name":"LayoutView","align_content":null,"align_items":null,"align_self":null,"border":null,"bottom":null,"display":null,"flex":null,"flex_flow":null,"grid_area":null,"grid_auto_columns":null,"grid_auto_flow":null,"grid_auto_rows":null,"grid_column":null,"grid_gap":null,"grid_row":null,"grid_template_areas":null,"grid_template_columns":null,"grid_template_rows":null,"height":null,"justify_content":null,"justify_items":null,"left":null,"margin":null,"max_height":null,"max_width":null,"min_height":null,"min_width":null,"object_fit":null,"object_position":null,"order":null,"overflow":null,"overflow_x":null,"overflow_y":null,"padding":null,"right":null,"top":null,"visibility":null,"width":null}},"6d1c7f3a39794b70bc22b0f823d53825":{"model_module":"@jupyter-widgets/controls","model_name":"DescriptionStyleModel","model_module_version":"1.5.0","state":{"_model_module":"@jupyter-widgets/controls","_model_module_version":"1.5.0","_model_name":"DescriptionStyleModel","_view_count":null,"_view_module":"@jupyter-widgets/base","_view_module_version":"1.2.0","_view_name":"StyleView","description_width":""}},"50049a3cc27146158d3f70c37e297568":{"model_module":"@jupyter-widgets/base","model_name":"LayoutModel","model_module_version":"1.2.0","state":{"_model_module":"@jupyter-widgets/base","_model_module_version":"1.2.0","_model_name":"LayoutModel","_view_count":null,"_view_module":"@jupyter-widgets/base","_view_module_version":"1.2.0","_view_name":"LayoutView","align_content":null,"align_items":null,"align_self":null,"border":null,"bottom":null,"display":null,"flex":null,"flex_flow":null,"grid_area":null,"grid_auto_columns":null,"grid_auto_flow":null,"grid_auto_rows":null,"grid_column":null,"grid_gap":null,"grid_row":null,"grid_template_areas":null,"grid_template_columns":null,"grid_template_rows":null,"height":null,"justify_content":null,"justify_items":null,"left":null,"margin":null,"max_height":null,"max_width":null,"min_height":null,"min_width":null,"object_fit":null,"object_position":null,"order":null,"overflow":null,"overflow_x":null,"overflow_y":null,"padding":null,"right":null,"top":null,"visibility":null,"width":null}},"5ad7fdb1a757485988c0ae1082663997":{"model_module":"@jupyter-widgets/controls","model_name":"ProgressStyleModel","model_module_version":"1.5.0","state":{"_model_module":"@jupyter-widgets/controls","_model_module_version":"1.5.0","_model_name":"ProgressStyleModel","_view_count":null,"_view_module":"@jupyter-widgets/base","_view_module_version":"1.2.0","_view_name":"StyleView","bar_color":null,"description_width":""}},"7f9a1321cb4a4882854ad29e7d2c4fbc":{"model_module":"@jupyter-widgets/base","model_name":"LayoutModel","model_module_version":"1.2.0","state":{"_model_module":"@jupyter-widgets/base","_model_module_version":"1.2.0","_model_name":"LayoutModel","_view_count":null,"_view_module":"@jupyter-widgets/base","_view_module_version":"1.2.0","_view_name":"LayoutView","align_content":null,"align_items":null,"align_self":null,"border":null,"bottom":null,"display":null,"flex":null,"flex_flow":null,"grid_area":null,"grid_auto_columns":null,"grid_auto_flow":null,"grid_auto_rows":null,"grid_column":null,"grid_gap":null,"grid_row":null,"grid_template_areas":null,"grid_template_columns":null,"grid_template_rows":null,"height":null,"justify_content":null,"justify_items":null,"left":null,"margin":null,"max_height":null,"max_width":null,"min_height":null,"min_width":null,"object_fit":null,"object_position":null,"order":null,"overflow":null,"overflow_x":null,"overflow_y":null,"padding":null,"right":null,"top":null,"visibility":null,"width":null}},"40e3ded372eb4f268ebc13c6b2c3b26c":{"model_module":"@jupyter-widgets/controls","model_name":"DescriptionStyleModel","model_module_version":"1.5.0","state":{"_model_module":"@jupyter-widgets/controls","_model_module_version":"1.5.0","_model_name":"DescriptionStyleModel","_view_count":null,"_view_module":"@jupyter-widgets/base","_view_module_version":"1.2.0","_view_name":"StyleView","description_width":""}}}}},"cells":[{"cell_type":"markdown","source":["# RAG Assesment"],"metadata":{"id":"OkMphBBDD4Ba"}},{"cell_type":"code","execution_count":1,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"MFpJ8RvAFbCM","executionInfo":{"status":"ok","timestamp":1756825572182,"user_tz":180,"elapsed":17950,"user":{"displayName":"Eugenia Sol Piñeiro","userId":"13178313008986059735"}},"outputId":"3f30e998-e47a-443a-d670-bfd85ee8c4cf","collapsed":true},"outputs":[{"output_type":"stream","name":"stdout","text":["Collecting ragas==0.1.9\n","  Downloading ragas-0.1.9-py3-none-any.whl.metadata (5.2 kB)\n","Requirement already satisfied: python-dotenv in /usr/local/lib/python3.12/dist-packages (1.1.1)\n","Requirement already satisfied: numpy in /usr/local/lib/python3.12/dist-packages (from ragas==0.1.9) (2.0.2)\n","Requirement already satisfied: datasets in /usr/local/lib/python3.12/dist-packages (from ragas==0.1.9) (4.0.0)\n","Requirement already satisfied: tiktoken in /usr/local/lib/python3.12/dist-packages (from ragas==0.1.9) (0.11.0)\n","Requirement already satisfied: langchain in /usr/local/lib/python3.12/dist-packages (from ragas==0.1.9) (0.3.27)\n","Requirement already satisfied: langchain-core in /usr/local/lib/python3.12/dist-packages (from ragas==0.1.9) (0.3.74)\n","Collecting langchain-community (from ragas==0.1.9)\n","  Downloading langchain_community-0.3.29-py3-none-any.whl.metadata (2.9 kB)\n","Collecting langchain-openai (from ragas==0.1.9)\n","  Downloading langchain_openai-0.3.32-py3-none-any.whl.metadata (2.4 kB)\n","Requirement already satisfied: openai>1 in /usr/local/lib/python3.12/dist-packages (from ragas==0.1.9) (1.101.0)\n","Collecting pysbd>=0.3.4 (from ragas==0.1.9)\n","  Downloading pysbd-0.3.4-py3-none-any.whl.metadata (6.1 kB)\n","Requirement already satisfied: nest-asyncio in /usr/local/lib/python3.12/dist-packages (from ragas==0.1.9) (1.6.0)\n","Collecting appdirs (from ragas==0.1.9)\n","  Downloading appdirs-1.4.4-py2.py3-none-any.whl.metadata (9.0 kB)\n","Requirement already satisfied: anyio<5,>=3.5.0 in /usr/local/lib/python3.12/dist-packages (from openai>1->ragas==0.1.9) (4.10.0)\n","Requirement already satisfied: distro<2,>=1.7.0 in /usr/local/lib/python3.12/dist-packages (from openai>1->ragas==0.1.9) (1.9.0)\n","Requirement already satisfied: httpx<1,>=0.23.0 in /usr/local/lib/python3.12/dist-packages (from openai>1->ragas==0.1.9) (0.28.1)\n","Requirement already satisfied: jiter<1,>=0.4.0 in /usr/local/lib/python3.12/dist-packages (from openai>1->ragas==0.1.9) (0.10.0)\n","Requirement already satisfied: pydantic<3,>=1.9.0 in /usr/local/lib/python3.12/dist-packages (from openai>1->ragas==0.1.9) (2.11.7)\n","Requirement already satisfied: sniffio in /usr/local/lib/python3.12/dist-packages (from openai>1->ragas==0.1.9) (1.3.1)\n","Requirement already satisfied: tqdm>4 in /usr/local/lib/python3.12/dist-packages (from openai>1->ragas==0.1.9) (4.67.1)\n","Requirement already satisfied: typing-extensions<5,>=4.11 in /usr/local/lib/python3.12/dist-packages (from openai>1->ragas==0.1.9) (4.15.0)\n","Requirement already satisfied: filelock in /usr/local/lib/python3.12/dist-packages (from datasets->ragas==0.1.9) (3.19.1)\n","Requirement already satisfied: pyarrow>=15.0.0 in /usr/local/lib/python3.12/dist-packages (from datasets->ragas==0.1.9) (18.1.0)\n","Requirement already satisfied: dill<0.3.9,>=0.3.0 in /usr/local/lib/python3.12/dist-packages (from datasets->ragas==0.1.9) (0.3.8)\n","Requirement already satisfied: pandas in /usr/local/lib/python3.12/dist-packages (from datasets->ragas==0.1.9) (2.2.2)\n","Requirement already satisfied: requests>=2.32.2 in /usr/local/lib/python3.12/dist-packages (from datasets->ragas==0.1.9) (2.32.4)\n","Requirement already satisfied: xxhash in /usr/local/lib/python3.12/dist-packages (from datasets->ragas==0.1.9) (3.5.0)\n","Requirement already satisfied: multiprocess<0.70.17 in /usr/local/lib/python3.12/dist-packages (from datasets->ragas==0.1.9) (0.70.16)\n","Requirement already satisfied: fsspec<=2025.3.0,>=2023.1.0 in /usr/local/lib/python3.12/dist-packages (from fsspec[http]<=2025.3.0,>=2023.1.0->datasets->ragas==0.1.9) (2025.3.0)\n","Requirement already satisfied: huggingface-hub>=0.24.0 in /usr/local/lib/python3.12/dist-packages (from datasets->ragas==0.1.9) (0.34.4)\n","Requirement already satisfied: packaging in /usr/local/lib/python3.12/dist-packages (from datasets->ragas==0.1.9) (25.0)\n","Requirement already satisfied: pyyaml>=5.1 in /usr/local/lib/python3.12/dist-packages (from datasets->ragas==0.1.9) (6.0.2)\n","Requirement already satisfied: langchain-text-splitters<1.0.0,>=0.3.9 in /usr/local/lib/python3.12/dist-packages (from langchain->ragas==0.1.9) (0.3.9)\n","Requirement already satisfied: langsmith>=0.1.17 in /usr/local/lib/python3.12/dist-packages (from langchain->ragas==0.1.9) (0.4.16)\n","Requirement already satisfied: SQLAlchemy<3,>=1.4 in /usr/local/lib/python3.12/dist-packages (from langchain->ragas==0.1.9) (2.0.43)\n","Requirement already satisfied: tenacity!=8.4.0,<10.0.0,>=8.1.0 in /usr/local/lib/python3.12/dist-packages (from langchain-core->ragas==0.1.9) (8.5.0)\n","Requirement already satisfied: jsonpatch<2.0,>=1.33 in /usr/local/lib/python3.12/dist-packages (from langchain-core->ragas==0.1.9) (1.33)\n","Collecting langchain-core (from ragas==0.1.9)\n","  Downloading langchain_core-0.3.75-py3-none-any.whl.metadata (5.7 kB)\n","Collecting requests>=2.32.2 (from datasets->ragas==0.1.9)\n","  Downloading requests-2.32.5-py3-none-any.whl.metadata (4.9 kB)\n","Requirement already satisfied: aiohttp<4.0.0,>=3.8.3 in /usr/local/lib/python3.12/dist-packages (from langchain-community->ragas==0.1.9) (3.12.15)\n","Collecting dataclasses-json<0.7,>=0.6.7 (from langchain-community->ragas==0.1.9)\n","  Downloading dataclasses_json-0.6.7-py3-none-any.whl.metadata (25 kB)\n","Requirement already satisfied: pydantic-settings<3.0.0,>=2.10.1 in /usr/local/lib/python3.12/dist-packages (from langchain-community->ragas==0.1.9) (2.10.1)\n","Requirement already satisfied: httpx-sse<1.0.0,>=0.4.0 in /usr/local/lib/python3.12/dist-packages (from langchain-community->ragas==0.1.9) (0.4.1)\n","Requirement already satisfied: regex>=2022.1.18 in /usr/local/lib/python3.12/dist-packages (from tiktoken->ragas==0.1.9) (2024.11.6)\n","Requirement already satisfied: aiohappyeyeballs>=2.5.0 in /usr/local/lib/python3.12/dist-packages (from aiohttp<4.0.0,>=3.8.3->langchain-community->ragas==0.1.9) (2.6.1)\n","Requirement already satisfied: aiosignal>=1.4.0 in /usr/local/lib/python3.12/dist-packages (from aiohttp<4.0.0,>=3.8.3->langchain-community->ragas==0.1.9) (1.4.0)\n","Requirement already satisfied: attrs>=17.3.0 in /usr/local/lib/python3.12/dist-packages (from aiohttp<4.0.0,>=3.8.3->langchain-community->ragas==0.1.9) (25.3.0)\n","Requirement already satisfied: frozenlist>=1.1.1 in /usr/local/lib/python3.12/dist-packages (from aiohttp<4.0.0,>=3.8.3->langchain-community->ragas==0.1.9) (1.7.0)\n","Requirement already satisfied: multidict<7.0,>=4.5 in /usr/local/lib/python3.12/dist-packages (from aiohttp<4.0.0,>=3.8.3->langchain-community->ragas==0.1.9) (6.6.4)\n","Requirement already satisfied: propcache>=0.2.0 in /usr/local/lib/python3.12/dist-packages (from aiohttp<4.0.0,>=3.8.3->langchain-community->ragas==0.1.9) (0.3.2)\n","Requirement already satisfied: yarl<2.0,>=1.17.0 in /usr/local/lib/python3.12/dist-packages (from aiohttp<4.0.0,>=3.8.3->langchain-community->ragas==0.1.9) (1.20.1)\n","Requirement already satisfied: idna>=2.8 in /usr/local/lib/python3.12/dist-packages (from anyio<5,>=3.5.0->openai>1->ragas==0.1.9) (3.10)\n","Collecting marshmallow<4.0.0,>=3.18.0 (from dataclasses-json<0.7,>=0.6.7->langchain-community->ragas==0.1.9)\n","  Downloading marshmallow-3.26.1-py3-none-any.whl.metadata (7.3 kB)\n","Collecting typing-inspect<1,>=0.4.0 (from dataclasses-json<0.7,>=0.6.7->langchain-community->ragas==0.1.9)\n","  Downloading typing_inspect-0.9.0-py3-none-any.whl.metadata (1.5 kB)\n","Requirement already satisfied: certifi in /usr/local/lib/python3.12/dist-packages (from httpx<1,>=0.23.0->openai>1->ragas==0.1.9) (2025.8.3)\n","Requirement already satisfied: httpcore==1.* in /usr/local/lib/python3.12/dist-packages (from httpx<1,>=0.23.0->openai>1->ragas==0.1.9) (1.0.9)\n","Requirement already satisfied: h11>=0.16 in /usr/local/lib/python3.12/dist-packages (from httpcore==1.*->httpx<1,>=0.23.0->openai>1->ragas==0.1.9) (0.16.0)\n","Requirement already satisfied: hf-xet<2.0.0,>=1.1.3 in /usr/local/lib/python3.12/dist-packages (from huggingface-hub>=0.24.0->datasets->ragas==0.1.9) (1.1.8)\n","Requirement already satisfied: jsonpointer>=1.9 in /usr/local/lib/python3.12/dist-packages (from jsonpatch<2.0,>=1.33->langchain-core->ragas==0.1.9) (3.0.0)\n","Requirement already satisfied: orjson>=3.9.14 in /usr/local/lib/python3.12/dist-packages (from langsmith>=0.1.17->langchain->ragas==0.1.9) (3.11.2)\n","Requirement already satisfied: requests-toolbelt>=1.0.0 in /usr/local/lib/python3.12/dist-packages (from langsmith>=0.1.17->langchain->ragas==0.1.9) (1.0.0)\n","Requirement already satisfied: zstandard>=0.23.0 in /usr/local/lib/python3.12/dist-packages (from langsmith>=0.1.17->langchain->ragas==0.1.9) (0.24.0)\n","Requirement already satisfied: annotated-types>=0.6.0 in /usr/local/lib/python3.12/dist-packages (from pydantic<3,>=1.9.0->openai>1->ragas==0.1.9) (0.7.0)\n","Requirement already satisfied: pydantic-core==2.33.2 in /usr/local/lib/python3.12/dist-packages (from pydantic<3,>=1.9.0->openai>1->ragas==0.1.9) (2.33.2)\n","Requirement already satisfied: typing-inspection>=0.4.0 in /usr/local/lib/python3.12/dist-packages (from pydantic<3,>=1.9.0->openai>1->ragas==0.1.9) (0.4.1)\n","Requirement already satisfied: charset_normalizer<4,>=2 in /usr/local/lib/python3.12/dist-packages (from requests>=2.32.2->datasets->ragas==0.1.9) (3.4.3)\n","Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.12/dist-packages (from requests>=2.32.2->datasets->ragas==0.1.9) (2.5.0)\n","Requirement already satisfied: greenlet>=1 in /usr/local/lib/python3.12/dist-packages (from SQLAlchemy<3,>=1.4->langchain->ragas==0.1.9) (3.2.4)\n","Requirement already satisfied: python-dateutil>=2.8.2 in /usr/local/lib/python3.12/dist-packages (from pandas->datasets->ragas==0.1.9) (2.9.0.post0)\n","Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.12/dist-packages (from pandas->datasets->ragas==0.1.9) (2025.2)\n","Requirement already satisfied: tzdata>=2022.7 in /usr/local/lib/python3.12/dist-packages (from pandas->datasets->ragas==0.1.9) (2025.2)\n","Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.12/dist-packages (from python-dateutil>=2.8.2->pandas->datasets->ragas==0.1.9) (1.17.0)\n","Collecting mypy-extensions>=0.3.0 (from typing-inspect<1,>=0.4.0->dataclasses-json<0.7,>=0.6.7->langchain-community->ragas==0.1.9)\n","  Downloading mypy_extensions-1.1.0-py3-none-any.whl.metadata (1.1 kB)\n","Downloading ragas-0.1.9-py3-none-any.whl (86 kB)\n","\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m86.1/86.1 kB\u001b[0m \u001b[31m3.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hDownloading pysbd-0.3.4-py3-none-any.whl (71 kB)\n","\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m71.1/71.1 kB\u001b[0m \u001b[31m5.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hDownloading appdirs-1.4.4-py2.py3-none-any.whl (9.6 kB)\n","Downloading langchain_community-0.3.29-py3-none-any.whl (2.5 MB)\n","\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m2.5/2.5 MB\u001b[0m \u001b[31m26.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hDownloading langchain_core-0.3.75-py3-none-any.whl (443 kB)\n","\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m444.0/444.0 kB\u001b[0m \u001b[31m15.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hDownloading langchain_openai-0.3.32-py3-none-any.whl (74 kB)\n","\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m74.5/74.5 kB\u001b[0m \u001b[31m5.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hDownloading dataclasses_json-0.6.7-py3-none-any.whl (28 kB)\n","Downloading requests-2.32.5-py3-none-any.whl (64 kB)\n","\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m64.7/64.7 kB\u001b[0m \u001b[31m4.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hDownloading marshmallow-3.26.1-py3-none-any.whl (50 kB)\n","\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m50.9/50.9 kB\u001b[0m \u001b[31m3.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hDownloading typing_inspect-0.9.0-py3-none-any.whl (8.8 kB)\n","Downloading mypy_extensions-1.1.0-py3-none-any.whl (5.0 kB)\n","Installing collected packages: appdirs, requests, pysbd, mypy-extensions, marshmallow, typing-inspect, dataclasses-json, langchain-core, langchain-openai, langchain-community, ragas\n","  Attempting uninstall: requests\n","    Found existing installation: requests 2.32.4\n","    Uninstalling requests-2.32.4:\n","      Successfully uninstalled requests-2.32.4\n","  Attempting uninstall: langchain-core\n","    Found existing installation: langchain-core 0.3.74\n","    Uninstalling langchain-core-0.3.74:\n","      Successfully uninstalled langchain-core-0.3.74\n","\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n","google-colab 1.0.0 requires requests==2.32.4, but you have requests 2.32.5 which is incompatible.\u001b[0m\u001b[31m\n","\u001b[0mSuccessfully installed appdirs-1.4.4 dataclasses-json-0.6.7 langchain-community-0.3.29 langchain-core-0.3.75 langchain-openai-0.3.32 marshmallow-3.26.1 mypy-extensions-1.1.0 pysbd-0.3.4 ragas-0.1.9 requests-2.32.5 typing-inspect-0.9.0\n"]}],"source":["pip install ragas==0.1.9 python-dotenv"]},{"cell_type":"code","source":["import os, openai\n","from dotenv import load_dotenv, find_dotenv\n","_ = load_dotenv(find_dotenv())"],"metadata":{"id":"PfGnfl80GC7p","executionInfo":{"status":"ok","timestamp":1756825573603,"user_tz":180,"elapsed":1427,"user":{"displayName":"Eugenia Sol Piñeiro","userId":"13178313008986059735"}}},"execution_count":2,"outputs":[]},{"cell_type":"code","source":["from datasets import Dataset\n","from ragas.metrics import faithfulness, answer_relevancy, context_precision, context_recall\n","from ragas import evaluate\n","\n","data_samples = {\n","    'question': ['When was Attention Is All You Need published and by whom?', 'When was Attention Is All You Need  published and by whom?'],\n","    'answer': ['Attention Is All You Need was published by Taylor Swift in 2017', 'Attention Is All You Need was published by Taylor Swift in 2003'],\n","    'contexts' : [['Attention Is All You Need was published by Vaswani et al in 2017'], ['Attention Is All You Need was published by Vaswani et al in 2017']],\n","}\n","dataset = Dataset.from_dict(data_samples)\n"],"metadata":{"id":"O_8eQX8OFpBz","executionInfo":{"status":"ok","timestamp":1756825584241,"user_tz":180,"elapsed":7662,"user":{"displayName":"Eugenia Sol Piñeiro","userId":"13178313008986059735"}},"colab":{"base_uri":"https://localhost:8080/"},"outputId":"1348468d-6e19-4cac-9fbb-9eab05dc620c"},"execution_count":3,"outputs":[{"output_type":"stream","name":"stderr","text":["/usr/local/lib/python3.12/dist-packages/ragas/metrics/__init__.py:1: LangChainDeprecationWarning: As of langchain-core 0.3.0, LangChain uses pydantic v2 internally. The langchain_core.pydantic_v1 module was a compatibility shim for pydantic v1, and should no longer be used. Please update the code to import from Pydantic directly.\n","\n","For example, replace imports like: `from langchain_core.pydantic_v1 import BaseModel`\n","with: `from pydantic import BaseModel`\n","or the v1 compatibility namespace if you are working in a code base that has not been fully upgraded to pydantic 2 yet. \tfrom pydantic.v1 import BaseModel\n","\n","  from ragas.metrics._answer_correctness import AnswerCorrectness, answer_correctness\n","/usr/local/lib/python3.12/dist-packages/ragas/metrics/__init__.py:4: LangChainDeprecationWarning: As of langchain-core 0.3.0, LangChain uses pydantic v2 internally. The langchain.pydantic_v1 module was a compatibility shim for pydantic v1, and should no longer be used. Please update the code to import from Pydantic directly.\n","\n","For example, replace imports like: `from langchain.pydantic_v1 import BaseModel`\n","with: `from pydantic import BaseModel`\n","or the v1 compatibility namespace if you are working in a code base that has not been fully upgraded to pydantic 2 yet. \tfrom pydantic.v1 import BaseModel\n","\n","  from ragas.metrics._context_entities_recall import (\n"]}]},{"cell_type":"code","source":["score = evaluate(dataset, metrics=[faithfulness,answer_relevancy],raise_exceptions=False)\n","score.to_pandas()"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":1000,"referenced_widgets":["21a6391881964522adcb3879ea507e9c","6214dfffeb5c4ec7a8f9f9182e9e7603","39ffd7cecf664f5ebe61fa8047b90d44","838d93b78b7c4249afe9c761886ccc86","a18d2efb1ef24c8e9060bc714b288ef0","2732175248a1418a9fab0f0cd845ecb7","6d1c7f3a39794b70bc22b0f823d53825","50049a3cc27146158d3f70c37e297568","5ad7fdb1a757485988c0ae1082663997","7f9a1321cb4a4882854ad29e7d2c4fbc","40e3ded372eb4f268ebc13c6b2c3b26c"]},"collapsed":true,"id":"4HTxCs0YGrlj","executionInfo":{"status":"ok","timestamp":1756825812324,"user_tz":180,"elapsed":185388,"user":{"displayName":"Eugenia Sol Piñeiro","userId":"13178313008986059735"}},"outputId":"efa39e26-95be-4a37-d8ec-0012babda234"},"execution_count":7,"outputs":[{"output_type":"display_data","data":{"text/plain":["Evaluating:   0%|          | 0/4 [00:00<?, ?it/s]"],"application/vnd.jupyter.widget-view+json":{"version_major":2,"version_minor":0,"model_id":"21a6391881964522adcb3879ea507e9c"}},"metadata":{}},{"output_type":"stream","name":"stderr","text":["ERROR:ragas.executor:Runner in Executor raised an exception\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.12/dist-packages/openai/_base_client.py\", line 1529, in request\n","    response = await self._client.send(\n","               ^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/httpx/_client.py\", line 1629, in send\n","    response = await self._send_handling_auth(\n","               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/httpx/_client.py\", line 1657, in _send_handling_auth\n","    response = await self._send_handling_redirects(\n","               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/httpx/_client.py\", line 1694, in _send_handling_redirects\n","    response = await self._send_single_request(request)\n","               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/httpx/_client.py\", line 1730, in _send_single_request\n","    response = await transport.handle_async_request(request)\n","               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/httpx/_transports/default.py\", line 394, in handle_async_request\n","    resp = await self._pool.handle_async_request(req)\n","           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/httpcore/_async/connection_pool.py\", line 256, in handle_async_request\n","    raise exc from None\n","  File \"/usr/local/lib/python3.12/dist-packages/httpcore/_async/connection_pool.py\", line 236, in handle_async_request\n","    response = await connection.handle_async_request(\n","               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/httpcore/_async/connection.py\", line 103, in handle_async_request\n","    return await self._connection.handle_async_request(request)\n","           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/httpcore/_async/http11.py\", line 136, in handle_async_request\n","    raise exc\n","  File \"/usr/local/lib/python3.12/dist-packages/httpcore/_async/http11.py\", line 106, in handle_async_request\n","    ) = await self._receive_response_headers(**kwargs)\n","        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/httpcore/_async/http11.py\", line 177, in _receive_response_headers\n","    event = await self._receive_event(timeout=timeout)\n","            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/httpcore/_async/http11.py\", line 217, in _receive_event\n","    data = await self._network_stream.read(\n","           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/httpcore/_backends/anyio.py\", line 35, in read\n","    return await self._stream.receive(max_bytes=max_bytes)\n","           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/anyio/streams/tls.py\", line 237, in receive\n","    data = await self._call_sslobject_method(self._ssl_object.read, max_bytes)\n","           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/anyio/streams/tls.py\", line 180, in _call_sslobject_method\n","    data = await self.transport_stream.receive()\n","           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/anyio/_backends/_asyncio.py\", line 1254, in receive\n","    await self._protocol.read_event.wait()\n","  File \"/usr/lib/python3.12/asyncio/locks.py\", line 209, in wait\n","    fut = self._get_loop().create_future()\n","          ^^^^^^^^^^^^^^^^\n","  File \"/usr/lib/python3.12/asyncio/mixins.py\", line 20, in _get_loop\n","    raise RuntimeError(f'{self!r} is bound to a different event loop')\n","RuntimeError: <asyncio.locks.Event object at 0x7cd4c66b46b0 [unset]> is bound to a different event loop\n","\n","The above exception was the direct cause of the following exception:\n","\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/executor.py\", line 78, in _aresults\n","    r = await future\n","        ^^^^^^^^^^^^\n","  File \"/usr/lib/python3.12/asyncio/tasks.py\", line 631, in _wait_for_one\n","    return f.result()  # May raise f.exception().\n","           ^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/executor.py\", line 37, in sema_coro\n","    return await coro\n","           ^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/executor.py\", line 111, in wrapped_callable_async\n","    return counter, await callable(*args, **kwargs)\n","                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/metrics/base.py\", line 125, in ascore\n","    raise e\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/metrics/base.py\", line 121, in ascore\n","    score = await self._ascore(row=row, callbacks=group_cm, is_async=is_async)\n","            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/metrics/_faithfulness.py\", line 248, in _ascore\n","    statements = await self.llm.generate(\n","                 ^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/llms/base.py\", line 93, in generate\n","    return await agenerate_text_with_retry(\n","           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/tenacity/asyncio/__init__.py\", line 189, in async_wrapped\n","    return await copy(fn, *args, **kwargs)\n","           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/tenacity/asyncio/__init__.py\", line 111, in __call__\n","    do = await self.iter(retry_state=retry_state)\n","         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/tenacity/asyncio/__init__.py\", line 153, in iter\n","    result = await action(retry_state)\n","             ^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/tenacity/_utils.py\", line 99, in inner\n","    return call(*args, **kwargs)\n","           ^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/tenacity/__init__.py\", line 398, in <lambda>\n","    self._add_action_func(lambda rs: rs.outcome.result())\n","                                     ^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/lib/python3.12/concurrent/futures/_base.py\", line 449, in result\n","    return self.__get_result()\n","           ^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/lib/python3.12/concurrent/futures/_base.py\", line 401, in __get_result\n","    raise self._exception\n","  File \"/usr/local/lib/python3.12/dist-packages/tenacity/asyncio/__init__.py\", line 114, in __call__\n","    result = await fn(*args, **kwargs)\n","             ^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/llms/base.py\", line 170, in agenerate_text\n","    return await self.langchain_llm.agenerate_prompt(\n","           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/langchain_core/language_models/chat_models.py\", line 1030, in agenerate_prompt\n","    return await self.agenerate(\n","           ^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/langchain_core/language_models/chat_models.py\", line 988, in agenerate\n","    raise exceptions[0]\n","  File \"/usr/local/lib/python3.12/dist-packages/langchain_core/language_models/chat_models.py\", line 1158, in _agenerate_with_cache\n","    result = await self._agenerate(\n","             ^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/langchain_openai/chat_models/base.py\", line 1429, in _agenerate\n","    raise e\n","  File \"/usr/local/lib/python3.12/dist-packages/langchain_openai/chat_models/base.py\", line 1422, in _agenerate\n","    raw_response = await self.async_client.with_raw_response.create(\n","                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/openai/_legacy_response.py\", line 381, in wrapped\n","    return cast(LegacyAPIResponse[R], await func(*args, **kwargs))\n","                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/openai/resources/chat/completions/completions.py\", line 2583, in create\n","    return await self._post(\n","           ^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/openai/_base_client.py\", line 1794, in post\n","    return await self.request(cast_to, opts, stream=stream, stream_cls=stream_cls)\n","           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/openai/_base_client.py\", line 1561, in request\n","    raise APIConnectionError(request=request) from err\n","openai.APIConnectionError: Connection error.\n","ERROR:ragas.executor:Runner in Executor raised an exception\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.12/dist-packages/openai/_base_client.py\", line 1529, in request\n","    response = await self._client.send(\n","               ^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/httpx/_client.py\", line 1629, in send\n","    response = await self._send_handling_auth(\n","               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/httpx/_client.py\", line 1657, in _send_handling_auth\n","    response = await self._send_handling_redirects(\n","               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/httpx/_client.py\", line 1694, in _send_handling_redirects\n","    response = await self._send_single_request(request)\n","               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/httpx/_client.py\", line 1730, in _send_single_request\n","    response = await transport.handle_async_request(request)\n","               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/httpx/_transports/default.py\", line 394, in handle_async_request\n","    resp = await self._pool.handle_async_request(req)\n","           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/httpcore/_async/connection_pool.py\", line 256, in handle_async_request\n","    raise exc from None\n","  File \"/usr/local/lib/python3.12/dist-packages/httpcore/_async/connection_pool.py\", line 236, in handle_async_request\n","    response = await connection.handle_async_request(\n","               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/httpcore/_async/connection.py\", line 103, in handle_async_request\n","    return await self._connection.handle_async_request(request)\n","           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/httpcore/_async/http11.py\", line 136, in handle_async_request\n","    raise exc\n","  File \"/usr/local/lib/python3.12/dist-packages/httpcore/_async/http11.py\", line 106, in handle_async_request\n","    ) = await self._receive_response_headers(**kwargs)\n","        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/httpcore/_async/http11.py\", line 177, in _receive_response_headers\n","    event = await self._receive_event(timeout=timeout)\n","            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/httpcore/_async/http11.py\", line 217, in _receive_event\n","    data = await self._network_stream.read(\n","           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/httpcore/_backends/anyio.py\", line 35, in read\n","    return await self._stream.receive(max_bytes=max_bytes)\n","           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/anyio/streams/tls.py\", line 237, in receive\n","    data = await self._call_sslobject_method(self._ssl_object.read, max_bytes)\n","           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/anyio/streams/tls.py\", line 180, in _call_sslobject_method\n","    data = await self.transport_stream.receive()\n","           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/anyio/_backends/_asyncio.py\", line 1254, in receive\n","    await self._protocol.read_event.wait()\n","  File \"/usr/lib/python3.12/asyncio/locks.py\", line 209, in wait\n","    fut = self._get_loop().create_future()\n","          ^^^^^^^^^^^^^^^^\n","  File \"/usr/lib/python3.12/asyncio/mixins.py\", line 20, in _get_loop\n","    raise RuntimeError(f'{self!r} is bound to a different event loop')\n","RuntimeError: <asyncio.locks.Event object at 0x7cd4c6781e80 [unset]> is bound to a different event loop\n","\n","The above exception was the direct cause of the following exception:\n","\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/executor.py\", line 78, in _aresults\n","    r = await future\n","        ^^^^^^^^^^^^\n","  File \"/usr/lib/python3.12/asyncio/tasks.py\", line 631, in _wait_for_one\n","    return f.result()  # May raise f.exception().\n","           ^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/executor.py\", line 37, in sema_coro\n","    return await coro\n","           ^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/executor.py\", line 111, in wrapped_callable_async\n","    return counter, await callable(*args, **kwargs)\n","                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/metrics/base.py\", line 125, in ascore\n","    raise e\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/metrics/base.py\", line 121, in ascore\n","    score = await self._ascore(row=row, callbacks=group_cm, is_async=is_async)\n","            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/metrics/_faithfulness.py\", line 248, in _ascore\n","    statements = await self.llm.generate(\n","                 ^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/llms/base.py\", line 93, in generate\n","    return await agenerate_text_with_retry(\n","           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/tenacity/asyncio/__init__.py\", line 189, in async_wrapped\n","    return await copy(fn, *args, **kwargs)\n","           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/tenacity/asyncio/__init__.py\", line 111, in __call__\n","    do = await self.iter(retry_state=retry_state)\n","         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/tenacity/asyncio/__init__.py\", line 153, in iter\n","    result = await action(retry_state)\n","             ^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/tenacity/_utils.py\", line 99, in inner\n","    return call(*args, **kwargs)\n","           ^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/tenacity/__init__.py\", line 398, in <lambda>\n","    self._add_action_func(lambda rs: rs.outcome.result())\n","                                     ^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/lib/python3.12/concurrent/futures/_base.py\", line 449, in result\n","    return self.__get_result()\n","           ^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/lib/python3.12/concurrent/futures/_base.py\", line 401, in __get_result\n","    raise self._exception\n","  File \"/usr/local/lib/python3.12/dist-packages/tenacity/asyncio/__init__.py\", line 114, in __call__\n","    result = await fn(*args, **kwargs)\n","             ^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/llms/base.py\", line 170, in agenerate_text\n","    return await self.langchain_llm.agenerate_prompt(\n","           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/langchain_core/language_models/chat_models.py\", line 1030, in agenerate_prompt\n","    return await self.agenerate(\n","           ^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/langchain_core/language_models/chat_models.py\", line 988, in agenerate\n","    raise exceptions[0]\n","  File \"/usr/local/lib/python3.12/dist-packages/langchain_core/language_models/chat_models.py\", line 1158, in _agenerate_with_cache\n","    result = await self._agenerate(\n","             ^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/langchain_openai/chat_models/base.py\", line 1429, in _agenerate\n","    raise e\n","  File \"/usr/local/lib/python3.12/dist-packages/langchain_openai/chat_models/base.py\", line 1422, in _agenerate\n","    raw_response = await self.async_client.with_raw_response.create(\n","                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/openai/_legacy_response.py\", line 381, in wrapped\n","    return cast(LegacyAPIResponse[R], await func(*args, **kwargs))\n","                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/openai/resources/chat/completions/completions.py\", line 2583, in create\n","    return await self._post(\n","           ^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/openai/_base_client.py\", line 1794, in post\n","    return await self.request(cast_to, opts, stream=stream, stream_cls=stream_cls)\n","           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/openai/_base_client.py\", line 1561, in request\n","    raise APIConnectionError(request=request) from err\n","openai.APIConnectionError: Connection error.\n","ERROR:ragas.executor:Runner in Executor raised an exception\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/executor.py\", line 78, in _aresults\n","    r = await future\n","        ^^^^^^^^^^^^\n","  File \"/usr/lib/python3.12/asyncio/tasks.py\", line 631, in _wait_for_one\n","    return f.result()  # May raise f.exception().\n","           ^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/executor.py\", line 37, in sema_coro\n","    return await coro\n","           ^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/executor.py\", line 111, in wrapped_callable_async\n","    return counter, await callable(*args, **kwargs)\n","                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/metrics/base.py\", line 125, in ascore\n","    raise e\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/metrics/base.py\", line 121, in ascore\n","    score = await self._ascore(row=row, callbacks=group_cm, is_async=is_async)\n","            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/metrics/_faithfulness.py\", line 248, in _ascore\n","    statements = await self.llm.generate(\n","                 ^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/llms/base.py\", line 93, in generate\n","    return await agenerate_text_with_retry(\n","           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/tenacity/asyncio/__init__.py\", line 189, in async_wrapped\n","    return await copy(fn, *args, **kwargs)\n","           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/tenacity/asyncio/__init__.py\", line 111, in __call__\n","    do = await self.iter(retry_state=retry_state)\n","         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/tenacity/asyncio/__init__.py\", line 153, in iter\n","    result = await action(retry_state)\n","             ^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/tenacity/_utils.py\", line 99, in inner\n","    return call(*args, **kwargs)\n","           ^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/tenacity/__init__.py\", line 418, in exc_check\n","    raise retry_exc.reraise()\n","          ^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/tenacity/__init__.py\", line 185, in reraise\n","    raise self.last_attempt.result()\n","          ^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/lib/python3.12/concurrent/futures/_base.py\", line 449, in result\n","    return self.__get_result()\n","           ^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/lib/python3.12/concurrent/futures/_base.py\", line 401, in __get_result\n","    raise self._exception\n","  File \"/usr/local/lib/python3.12/dist-packages/tenacity/asyncio/__init__.py\", line 114, in __call__\n","    result = await fn(*args, **kwargs)\n","             ^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/llms/base.py\", line 170, in agenerate_text\n","    return await self.langchain_llm.agenerate_prompt(\n","           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/langchain_core/language_models/chat_models.py\", line 1030, in agenerate_prompt\n","    return await self.agenerate(\n","           ^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/langchain_core/language_models/chat_models.py\", line 988, in agenerate\n","    raise exceptions[0]\n","  File \"/usr/local/lib/python3.12/dist-packages/langchain_core/language_models/chat_models.py\", line 1158, in _agenerate_with_cache\n","    result = await self._agenerate(\n","             ^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/langchain_openai/chat_models/base.py\", line 1429, in _agenerate\n","    raise e\n","  File \"/usr/local/lib/python3.12/dist-packages/langchain_openai/chat_models/base.py\", line 1422, in _agenerate\n","    raw_response = await self.async_client.with_raw_response.create(\n","                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/openai/_legacy_response.py\", line 381, in wrapped\n","    return cast(LegacyAPIResponse[R], await func(*args, **kwargs))\n","                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/openai/resources/chat/completions/completions.py\", line 2583, in create\n","    return await self._post(\n","           ^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/openai/_base_client.py\", line 1794, in post\n","    return await self.request(cast_to, opts, stream=stream, stream_cls=stream_cls)\n","           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/openai/_base_client.py\", line 1594, in request\n","    raise self._make_status_error_from_response(err.response) from None\n","openai.RateLimitError: Error code: 429 - {'error': {'message': 'You exceeded your current quota, please check your plan and billing details. For more information on this error, read the docs: https://platform.openai.com/docs/guides/error-codes/api-errors.', 'type': 'insufficient_quota', 'param': None, 'code': 'insufficient_quota'}}\n","ERROR:ragas.executor:Runner in Executor raised an exception\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/executor.py\", line 78, in _aresults\n","    r = await future\n","        ^^^^^^^^^^^^\n","  File \"/usr/lib/python3.12/asyncio/tasks.py\", line 631, in _wait_for_one\n","    return f.result()  # May raise f.exception().\n","           ^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/executor.py\", line 37, in sema_coro\n","    return await coro\n","           ^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/executor.py\", line 111, in wrapped_callable_async\n","    return counter, await callable(*args, **kwargs)\n","                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/metrics/base.py\", line 125, in ascore\n","    raise e\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/metrics/base.py\", line 121, in ascore\n","    score = await self._ascore(row=row, callbacks=group_cm, is_async=is_async)\n","            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/metrics/_answer_relevance.py\", line 152, in _ascore\n","    result = await self.llm.generate(\n","             ^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/llms/base.py\", line 93, in generate\n","    return await agenerate_text_with_retry(\n","           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/tenacity/asyncio/__init__.py\", line 189, in async_wrapped\n","    return await copy(fn, *args, **kwargs)\n","           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/tenacity/asyncio/__init__.py\", line 111, in __call__\n","    do = await self.iter(retry_state=retry_state)\n","         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/tenacity/asyncio/__init__.py\", line 153, in iter\n","    result = await action(retry_state)\n","             ^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/tenacity/_utils.py\", line 99, in inner\n","    return call(*args, **kwargs)\n","           ^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/tenacity/__init__.py\", line 418, in exc_check\n","    raise retry_exc.reraise()\n","          ^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/tenacity/__init__.py\", line 185, in reraise\n","    raise self.last_attempt.result()\n","          ^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/lib/python3.12/concurrent/futures/_base.py\", line 449, in result\n","    return self.__get_result()\n","           ^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/lib/python3.12/concurrent/futures/_base.py\", line 401, in __get_result\n","    raise self._exception\n","  File \"/usr/local/lib/python3.12/dist-packages/tenacity/asyncio/__init__.py\", line 114, in __call__\n","    result = await fn(*args, **kwargs)\n","             ^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/llms/base.py\", line 170, in agenerate_text\n","    return await self.langchain_llm.agenerate_prompt(\n","           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/langchain_core/language_models/chat_models.py\", line 1030, in agenerate_prompt\n","    return await self.agenerate(\n","           ^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/langchain_core/language_models/chat_models.py\", line 988, in agenerate\n","    raise exceptions[0]\n","  File \"/usr/local/lib/python3.12/dist-packages/langchain_core/language_models/chat_models.py\", line 1158, in _agenerate_with_cache\n","    result = await self._agenerate(\n","             ^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/langchain_openai/chat_models/base.py\", line 1429, in _agenerate\n","    raise e\n","  File \"/usr/local/lib/python3.12/dist-packages/langchain_openai/chat_models/base.py\", line 1422, in _agenerate\n","    raw_response = await self.async_client.with_raw_response.create(\n","                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/openai/_legacy_response.py\", line 381, in wrapped\n","    return cast(LegacyAPIResponse[R], await func(*args, **kwargs))\n","                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/openai/resources/chat/completions/completions.py\", line 2583, in create\n","    return await self._post(\n","           ^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/openai/_base_client.py\", line 1794, in post\n","    return await self.request(cast_to, opts, stream=stream, stream_cls=stream_cls)\n","           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/openai/_base_client.py\", line 1594, in request\n","    raise self._make_status_error_from_response(err.response) from None\n","openai.RateLimitError: Error code: 429 - {'error': {'message': 'You exceeded your current quota, please check your plan and billing details. For more information on this error, read the docs: https://platform.openai.com/docs/guides/error-codes/api-errors.', 'type': 'insufficient_quota', 'param': None, 'code': 'insufficient_quota'}}\n","ERROR:ragas.executor:Runner in Executor raised an exception\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/executor.py\", line 78, in _aresults\n","    r = await future\n","        ^^^^^^^^^^^^\n","  File \"/usr/lib/python3.12/asyncio/tasks.py\", line 631, in _wait_for_one\n","    return f.result()  # May raise f.exception().\n","           ^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/executor.py\", line 37, in sema_coro\n","    return await coro\n","           ^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/executor.py\", line 111, in wrapped_callable_async\n","    return counter, await callable(*args, **kwargs)\n","                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/metrics/base.py\", line 125, in ascore\n","    raise e\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/metrics/base.py\", line 121, in ascore\n","    score = await self._ascore(row=row, callbacks=group_cm, is_async=is_async)\n","            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/metrics/_faithfulness.py\", line 248, in _ascore\n","    statements = await self.llm.generate(\n","                 ^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/llms/base.py\", line 93, in generate\n","    return await agenerate_text_with_retry(\n","           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/tenacity/asyncio/__init__.py\", line 189, in async_wrapped\n","    return await copy(fn, *args, **kwargs)\n","           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/tenacity/asyncio/__init__.py\", line 111, in __call__\n","    do = await self.iter(retry_state=retry_state)\n","         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/tenacity/asyncio/__init__.py\", line 153, in iter\n","    result = await action(retry_state)\n","             ^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/tenacity/_utils.py\", line 99, in inner\n","    return call(*args, **kwargs)\n","           ^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/tenacity/__init__.py\", line 418, in exc_check\n","    raise retry_exc.reraise()\n","          ^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/tenacity/__init__.py\", line 185, in reraise\n","    raise self.last_attempt.result()\n","          ^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/lib/python3.12/concurrent/futures/_base.py\", line 449, in result\n","    return self.__get_result()\n","           ^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/lib/python3.12/concurrent/futures/_base.py\", line 401, in __get_result\n","    raise self._exception\n","  File \"/usr/local/lib/python3.12/dist-packages/tenacity/asyncio/__init__.py\", line 114, in __call__\n","    result = await fn(*args, **kwargs)\n","             ^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/llms/base.py\", line 170, in agenerate_text\n","    return await self.langchain_llm.agenerate_prompt(\n","           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/langchain_core/language_models/chat_models.py\", line 1030, in agenerate_prompt\n","    return await self.agenerate(\n","           ^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/langchain_core/language_models/chat_models.py\", line 988, in agenerate\n","    raise exceptions[0]\n","  File \"/usr/local/lib/python3.12/dist-packages/langchain_core/language_models/chat_models.py\", line 1158, in _agenerate_with_cache\n","    result = await self._agenerate(\n","             ^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/langchain_openai/chat_models/base.py\", line 1429, in _agenerate\n","    raise e\n","  File \"/usr/local/lib/python3.12/dist-packages/langchain_openai/chat_models/base.py\", line 1422, in _agenerate\n","    raw_response = await self.async_client.with_raw_response.create(\n","                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/openai/_legacy_response.py\", line 381, in wrapped\n","    return cast(LegacyAPIResponse[R], await func(*args, **kwargs))\n","                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/openai/resources/chat/completions/completions.py\", line 2583, in create\n","    return await self._post(\n","           ^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/openai/_base_client.py\", line 1794, in post\n","    return await self.request(cast_to, opts, stream=stream, stream_cls=stream_cls)\n","           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/openai/_base_client.py\", line 1594, in request\n","    raise self._make_status_error_from_response(err.response) from None\n","openai.RateLimitError: Error code: 429 - {'error': {'message': 'You exceeded your current quota, please check your plan and billing details. For more information on this error, read the docs: https://platform.openai.com/docs/guides/error-codes/api-errors.', 'type': 'insufficient_quota', 'param': None, 'code': 'insufficient_quota'}}\n","ERROR:ragas.executor:Runner in Executor raised an exception\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/executor.py\", line 78, in _aresults\n","    r = await future\n","        ^^^^^^^^^^^^\n","  File \"/usr/lib/python3.12/asyncio/tasks.py\", line 631, in _wait_for_one\n","    return f.result()  # May raise f.exception().\n","           ^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/executor.py\", line 37, in sema_coro\n","    return await coro\n","           ^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/executor.py\", line 111, in wrapped_callable_async\n","    return counter, await callable(*args, **kwargs)\n","                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/metrics/base.py\", line 125, in ascore\n","    raise e\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/metrics/base.py\", line 121, in ascore\n","    score = await self._ascore(row=row, callbacks=group_cm, is_async=is_async)\n","            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/metrics/_answer_relevance.py\", line 152, in _ascore\n","    result = await self.llm.generate(\n","             ^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/llms/base.py\", line 93, in generate\n","    return await agenerate_text_with_retry(\n","           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/tenacity/asyncio/__init__.py\", line 189, in async_wrapped\n","    return await copy(fn, *args, **kwargs)\n","           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/tenacity/asyncio/__init__.py\", line 111, in __call__\n","    do = await self.iter(retry_state=retry_state)\n","         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/tenacity/asyncio/__init__.py\", line 153, in iter\n","    result = await action(retry_state)\n","             ^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/tenacity/_utils.py\", line 99, in inner\n","    return call(*args, **kwargs)\n","           ^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/tenacity/__init__.py\", line 418, in exc_check\n","    raise retry_exc.reraise()\n","          ^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/tenacity/__init__.py\", line 185, in reraise\n","    raise self.last_attempt.result()\n","          ^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/lib/python3.12/concurrent/futures/_base.py\", line 449, in result\n","    return self.__get_result()\n","           ^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/lib/python3.12/concurrent/futures/_base.py\", line 401, in __get_result\n","    raise self._exception\n","  File \"/usr/local/lib/python3.12/dist-packages/tenacity/asyncio/__init__.py\", line 114, in __call__\n","    result = await fn(*args, **kwargs)\n","             ^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/ragas/llms/base.py\", line 170, in agenerate_text\n","    return await self.langchain_llm.agenerate_prompt(\n","           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/langchain_core/language_models/chat_models.py\", line 1030, in agenerate_prompt\n","    return await self.agenerate(\n","           ^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/langchain_core/language_models/chat_models.py\", line 988, in agenerate\n","    raise exceptions[0]\n","  File \"/usr/local/lib/python3.12/dist-packages/langchain_core/language_models/chat_models.py\", line 1158, in _agenerate_with_cache\n","    result = await self._agenerate(\n","             ^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/langchain_openai/chat_models/base.py\", line 1429, in _agenerate\n","    raise e\n","  File \"/usr/local/lib/python3.12/dist-packages/langchain_openai/chat_models/base.py\", line 1422, in _agenerate\n","    raw_response = await self.async_client.with_raw_response.create(\n","                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/openai/_legacy_response.py\", line 381, in wrapped\n","    return cast(LegacyAPIResponse[R], await func(*args, **kwargs))\n","                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/openai/resources/chat/completions/completions.py\", line 2583, in create\n","    return await self._post(\n","           ^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/openai/_base_client.py\", line 1794, in post\n","    return await self.request(cast_to, opts, stream=stream, stream_cls=stream_cls)\n","           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.12/dist-packages/openai/_base_client.py\", line 1594, in request\n","    raise self._make_status_error_from_response(err.response) from None\n","openai.RateLimitError: Error code: 429 - {'error': {'message': 'You exceeded your current quota, please check your plan and billing details. For more information on this error, read the docs: https://platform.openai.com/docs/guides/error-codes/api-errors.', 'type': 'insufficient_quota', 'param': None, 'code': 'insufficient_quota'}}\n","/usr/local/lib/python3.12/dist-packages/ragas/evaluation.py:299: RuntimeWarning: Mean of empty slice\n","  value = np.nanmean(self.scores[cn])\n"]},{"output_type":"execute_result","data":{"text/plain":["                                            question  \\\n","0  When was Attention Is All You Need published a...   \n","1  When was Attention Is All You Need  published ...   \n","\n","                                              answer  \\\n","0  Attention Is All You Need was published by Tay...   \n","1  Attention Is All You Need was published by Tay...   \n","\n","                                            contexts  faithfulness  \\\n","0  [Attention Is All You Need was published by Va...           NaN   \n","1  [Attention Is All You Need was published by Va...           NaN   \n","\n","   answer_relevancy  \n","0               NaN  \n","1               NaN  "],"text/html":["\n","  <div id=\"df-2bd8eb97-7c1b-46cb-a776-2a76f58b83e1\" class=\"colab-df-container\">\n","    <div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>question</th>\n","      <th>answer</th>\n","      <th>contexts</th>\n","      <th>faithfulness</th>\n","      <th>answer_relevancy</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>When was Attention Is All You Need published a...</td>\n","      <td>Attention Is All You Need was published by Tay...</td>\n","      <td>[Attention Is All You Need was published by Va...</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>When was Attention Is All You Need  published ...</td>\n","      <td>Attention Is All You Need was published by Tay...</td>\n","      <td>[Attention Is All You Need was published by Va...</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>\n","    <div class=\"colab-df-buttons\">\n","\n","  <div class=\"colab-df-container\">\n","    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-2bd8eb97-7c1b-46cb-a776-2a76f58b83e1')\"\n","            title=\"Convert this dataframe to an interactive table.\"\n","            style=\"display:none;\">\n","\n","  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n","    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n","  </svg>\n","    </button>\n","\n","  <style>\n","    .colab-df-container {\n","      display:flex;\n","      gap: 12px;\n","    }\n","\n","    .colab-df-convert {\n","      background-color: #E8F0FE;\n","      border: none;\n","      border-radius: 50%;\n","      cursor: pointer;\n","      display: none;\n","      fill: #1967D2;\n","      height: 32px;\n","      padding: 0 0 0 0;\n","      width: 32px;\n","    }\n","\n","    .colab-df-convert:hover {\n","      background-color: #E2EBFA;\n","      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n","      fill: #174EA6;\n","    }\n","\n","    .colab-df-buttons div {\n","      margin-bottom: 4px;\n","    }\n","\n","    [theme=dark] .colab-df-convert {\n","      background-color: #3B4455;\n","      fill: #D2E3FC;\n","    }\n","\n","    [theme=dark] .colab-df-convert:hover {\n","      background-color: #434B5C;\n","      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n","      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n","      fill: #FFFFFF;\n","    }\n","  </style>\n","\n","    <script>\n","      const buttonEl =\n","        document.querySelector('#df-2bd8eb97-7c1b-46cb-a776-2a76f58b83e1 button.colab-df-convert');\n","      buttonEl.style.display =\n","        google.colab.kernel.accessAllowed ? 'block' : 'none';\n","\n","      async function convertToInteractive(key) {\n","        const element = document.querySelector('#df-2bd8eb97-7c1b-46cb-a776-2a76f58b83e1');\n","        const dataTable =\n","          await google.colab.kernel.invokeFunction('convertToInteractive',\n","                                                    [key], {});\n","        if (!dataTable) return;\n","\n","        const docLinkHtml = 'Like what you see? Visit the ' +\n","          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n","          + ' to learn more about interactive tables.';\n","        element.innerHTML = '';\n","        dataTable['output_type'] = 'display_data';\n","        await google.colab.output.renderOutput(dataTable, element);\n","        const docLink = document.createElement('div');\n","        docLink.innerHTML = docLinkHtml;\n","        element.appendChild(docLink);\n","      }\n","    </script>\n","  </div>\n","\n","\n","    <div id=\"df-82ba227a-6a15-4baf-8bc4-261ac5bd5d69\">\n","      <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-82ba227a-6a15-4baf-8bc4-261ac5bd5d69')\"\n","                title=\"Suggest charts\"\n","                style=\"display:none;\">\n","\n","<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n","     width=\"24px\">\n","    <g>\n","        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n","    </g>\n","</svg>\n","      </button>\n","\n","<style>\n","  .colab-df-quickchart {\n","      --bg-color: #E8F0FE;\n","      --fill-color: #1967D2;\n","      --hover-bg-color: #E2EBFA;\n","      --hover-fill-color: #174EA6;\n","      --disabled-fill-color: #AAA;\n","      --disabled-bg-color: #DDD;\n","  }\n","\n","  [theme=dark] .colab-df-quickchart {\n","      --bg-color: #3B4455;\n","      --fill-color: #D2E3FC;\n","      --hover-bg-color: #434B5C;\n","      --hover-fill-color: #FFFFFF;\n","      --disabled-bg-color: #3B4455;\n","      --disabled-fill-color: #666;\n","  }\n","\n","  .colab-df-quickchart {\n","    background-color: var(--bg-color);\n","    border: none;\n","    border-radius: 50%;\n","    cursor: pointer;\n","    display: none;\n","    fill: var(--fill-color);\n","    height: 32px;\n","    padding: 0;\n","    width: 32px;\n","  }\n","\n","  .colab-df-quickchart:hover {\n","    background-color: var(--hover-bg-color);\n","    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n","    fill: var(--button-hover-fill-color);\n","  }\n","\n","  .colab-df-quickchart-complete:disabled,\n","  .colab-df-quickchart-complete:disabled:hover {\n","    background-color: var(--disabled-bg-color);\n","    fill: var(--disabled-fill-color);\n","    box-shadow: none;\n","  }\n","\n","  .colab-df-spinner {\n","    border: 2px solid var(--fill-color);\n","    border-color: transparent;\n","    border-bottom-color: var(--fill-color);\n","    animation:\n","      spin 1s steps(1) infinite;\n","  }\n","\n","  @keyframes spin {\n","    0% {\n","      border-color: transparent;\n","      border-bottom-color: var(--fill-color);\n","      border-left-color: var(--fill-color);\n","    }\n","    20% {\n","      border-color: transparent;\n","      border-left-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","    }\n","    30% {\n","      border-color: transparent;\n","      border-left-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","      border-right-color: var(--fill-color);\n","    }\n","    40% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","    }\n","    60% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","    }\n","    80% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","      border-bottom-color: var(--fill-color);\n","    }\n","    90% {\n","      border-color: transparent;\n","      border-bottom-color: var(--fill-color);\n","    }\n","  }\n","</style>\n","\n","      <script>\n","        async function quickchart(key) {\n","          const quickchartButtonEl =\n","            document.querySelector('#' + key + ' button');\n","          quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n","          quickchartButtonEl.classList.add('colab-df-spinner');\n","          try {\n","            const charts = await google.colab.kernel.invokeFunction(\n","                'suggestCharts', [key], {});\n","          } catch (error) {\n","            console.error('Error during call to suggestCharts:', error);\n","          }\n","          quickchartButtonEl.classList.remove('colab-df-spinner');\n","          quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n","        }\n","        (() => {\n","          let quickchartButtonEl =\n","            document.querySelector('#df-82ba227a-6a15-4baf-8bc4-261ac5bd5d69 button');\n","          quickchartButtonEl.style.display =\n","            google.colab.kernel.accessAllowed ? 'block' : 'none';\n","        })();\n","      </script>\n","    </div>\n","\n","    </div>\n","  </div>\n"],"application/vnd.google.colaboratory.intrinsic+json":{"type":"dataframe","summary":"{\n  \"name\": \"score\",\n  \"rows\": 2,\n  \"fields\": [\n    {\n      \"column\": \"question\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 2,\n        \"samples\": [\n          \"When was Attention Is All You Need  published and by whom?\",\n          \"When was Attention Is All You Need published and by whom?\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"answer\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 2,\n        \"samples\": [\n          \"Attention Is All You Need was published by Taylor Swift in 2003\",\n          \"Attention Is All You Need was published by Taylor Swift in 2017\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"contexts\",\n      \"properties\": {\n        \"dtype\": \"object\",\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"faithfulness\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": null,\n        \"min\": null,\n        \"max\": null,\n        \"num_unique_values\": 0,\n        \"samples\": [],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"answer_relevancy\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": null,\n        \"min\": null,\n        \"max\": null,\n        \"num_unique_values\": 0,\n        \"samples\": [],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"}},"metadata":{},"execution_count":7}]}]}